bson/src/main/org/bson/codecs/pojo/RecordCodecProvider.java

```diff
@@ -0,0 +1,202 @@
+import java.util.Map;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static org.bson.assertions.Assertions.notNull;
+
+/**
+ * Provides Codecs for registered Records via the ClassModel abstractions.
+ *
+ * @since TODO
+ */
+public final class RecordCodecProvider implements CodecProvider {
+    static final Logger LOGGER = Loggers.getLogger("codecs.record");
+    private final boolean automatic;
+    private final Map<Class<?>, ClassModel<?>> classModels;
+    private final Set<String> packages;
+    private final DiscriminatorLookup discriminatorLookup;
+    private final List<PropertyCodecProvider> propertyCodecProviders;
+
+    private RecordCodecProvider(final boolean automatic, final Map<Class<?>, ClassModel<?>> classModels, final Set<String> packages,
+                              final List<PropertyCodecProvider> propertyCodecProviders) {
+        this.automatic = automatic;
+        this.classModels = classModels;
+        this.packages = packages;
+        this.discriminatorLookup = new DiscriminatorLookup(classModels, packages);
+        this.propertyCodecProviders = propertyCodecProviders;
+    }
+
+    /**
+     * Creates a Builder so classes or packages can be registered and configured before creating an immutable CodecProvider.
+     *
+     * @return the Builder
+     * @see RecordCodecProvider.Builder#register(Class[])
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> Codec<T> get(final Class<T> clazz, final CodecRegistry registry) {
+        try {
+            clazz.getClass().getMethod("isRecord");
+            if (!clazz.isRecord()) {
+                return null;
+            }
+        } catch (NoSuchMethodException e) {
+            return null;
+        }
+
+        ClassModel<T> classModel = (ClassModel<T>) classModels.get(clazz);
+        if (classModel != null) {
+            return new PojoCodecImpl<T>(classModel, registry, propertyCodecProviders, discriminatorLookup);
+        } else if (automatic || (clazz.getPackage() != null && packages.contains(clazz.getPackage().getName()))) {
+            try {
+                classModel = createClassModel(clazz);
+                if (clazz.isInterface() || !classModel.getPropertyModels().isEmpty()) {
+                    discriminatorLookup.addClassModel(classModel);
+                    return new AutomaticPojoCodec<T>(new PojoCodecImpl<T>(classModel, registry, propertyCodecProviders,
+                            discriminatorLookup));
+                }
+            } catch (Exception e) {
+                LOGGER.warn(format("Cannot use '%s' with the PojoCodec for records.", clazz.getSimpleName()), e);
+                return null;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * A Builder for the RecordCodecProvider
+     */
+    public static final class Builder {
```

Comment:
This class unfortunately contains a lot of duplicate code with the Builder in PojoCodecProvider. I considered making an abstract builder, but that would require me to either make the instance variables protected (which the style checker complained about) or add getter methods for all of the instance variables. I figured the extra code that was required to make that work wasn't worth it, but let me know if that was a mistake.

Reply:
We also may want to consider making an abstract UserDefinedCodecProvider, because RecordCodecProvider and PojoCodecProvider are so similar.